src/clients/Deluge.ts

```diff
@@ -0,0 +1,232 @@
+import { InjectionResult } from "../constants.js";
+import { CrossSeedError } from "../errors.js";
+import { Label, logger } from "../logger.js";
+import { Metafile } from "../parseTorrent.js";
+import { getRuntimeConfig } from "../runtimeConfig.js";
+import { Searchee } from "../searchee.js";
+import { TorrentClient } from "./TorrentClient.js";
+import fetch, { Headers, Response } from "node-fetch";
+
+interface DelugeResponse {
+	error?: {
+		message?: string;
+		code?: number;
+	};
+	result?: any;
+}
+
+export default class Deluge implements TorrentClient {
+	private msgId = 0;
+	private delugeCookie = "";
+	private delugeWebUrl: URL;
+	private delugeLabel: string;
+	private isLabelEnabled: Promise<boolean>;
+
+	constructor() {
+		this.delugeWebUrl = new URL(`${getRuntimeConfig().delugeWebUrl}`);
+		this.delugeLabel = "cross-seed";
+	}
+
+	/**
+	 * validates the login and host for deluge webui
+	 */
+	async validateConfig(): Promise<void> {
+		const url = new URL(this.delugeWebUrl);
+		if (url.username && !url.password) {
+			throw new CrossSeedError(
+				"you need to define a password in the delugeWebUrl. (eg: http://:<PASSWORD>@localhost:8112)"
+			);
+		}
+		await this.authenticate();
+		this.isLabelEnabled = this.labelEnabled();
+	}
+
+	/**
+	 * connects and authenticates to the webui
+	 */
+	private async authenticate(): Promise<void> {
+		const response = await this.call({
+			method: "auth.login",
+			params: [this.delugeWebUrl.password],
+		});
+
+		if (response === null) {
+			throw new CrossSeedError(
+				`failed to establish a connection to deluge: ${this.delugeWebUrl.origin}`
+			);
+		}
+	}
+
+	/**
+	 * ensures authentication and sends JSON-RPC calls to deluge
+	 */
+	private async call(body: { method: string; params: any[] }) {
+		this.msgId = (this.msgId + 1) % 1024;
+
+		const headers = new Headers({ "Content-Type": "application/json" });
+		if (this.delugeCookie) headers.set("Cookie", this.delugeCookie);
+
+		const url = this.delugeWebUrl.origin + this.delugeWebUrl.pathname;
+		let response: Response, json: DelugeResponse;
+		try {
+			response = await fetch(url, {
+				body: JSON.stringify({ ...body, id: this.msgId }),
+				method: "POST",
+				headers: headers,
+			});
+			json = (await response.json()) as DelugeResponse;
+		} catch (networkError) {
+			return null;
+		}
+
+		if (json && json.error && json.error.code && json.error.code === 1) {
```

Comment:
We can use [Optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) syntax here (also elsewhere):
```suggestion
		if (json?.error?.code === 1) {
```